File: experiments/problem1_nonperiodicity/analyze_period_two_arithmetic_quotient.py
#!/usr/bin/env python3
"""Analyze the arithmetic quotient of normalized period-two witnesses.

For a normalized inverse word G=a_1...a_n beginning in phase p or u, let

    x(G)=G^{-1}(0)

be its ordinary preimage of zero. Different words can have the same x(G), and
on odd arithmetic states the forward letters p and u coincide. Tracking only
state parity gives a Pell-growth upper bound on the number of distinct
arithmetic witnesses:

    E_(n+1) <= E_n + O_n,
    O_(n+1) <= 2 E_n + O_n.

For one fixed phase, the number of exact-length-n states is at most the Pell
number P_n. This sharpens the finite-witness counting theorem and improves the
Bernoulli-almost-sure lower rate from log(2)/log(3) to
log(2)/log(1+sqrt(2)).

The bounded campaign also compares every reachable arithmetic state through a
configured word length with the exact actual schedule-survivor residue. It is
finite regression evidence and does not prove divergence on the actual orbit.
"""
from __future__ import annotations

import argparse
import hashlib
import json
import math
from typing import Any

LETTERS = ("t", "p", "u")
PHASE_START = {"p": 3, "u": 1}
PHASES = tuple(PHASE_START)
DEFAULT_MAXIMUM_LENGTH = 16
ABSOLUTE_MAXIMUM_LENGTH = 22
DEFAULT_SURVIVOR_BITS = 64


def forward_generator(name: str, state: int) -> int:
    if state < 0:
        raise ValueError("state must be nonnegative")
    stepped = state ^ ((state << 1) | (state << 2))
    if name == "t":
        return stepped
    if name == "u":
        return stepped ^ 1
    if name == "p":
        return stepped ^ 1 ^ (2 if state & 1 == 0 else 0)
    raise ValueError("unknown generator")


def next_arithmetic_states(state: int) -> set[int]:
    return {forward_generator(letter, state) for letter in LETTERS}


def pell_numbers(maximum_index: int) -> list[int]:
    if maximum_index < 0:
        raise ValueError("maximum index must be nonnegative")
    values = [0] * (maximum_index + 1)
    if maximum_index >= 1:
        values[1] = 1
    for index in range(2, maximum_index + 1):
        values[index] = 2 * values[index - 1] + values[index - 2]
    return values


def pell_cumulative_bound(maximum_length: int) -> int:
    """Return sum(P_1,...,P_N)."""
    if maximum_length <= 0:
        return 0
    pell = pell_numbers(maximum_length + 1)
    return (pell[maximum_length + 1] + pell[maximum_length] - 1) // 2


def verify_parity_transition(sample_limit: int = 4096) -> dict[str, Any]:
    checked = 0
    for state in range(1, sample_limit + 1):
        t_value = forward_generator("t", state)
        p_value = forward_generator("p", state)
        u_value = forward_generator("u", state)
        if (t_value & 1) != (state & 1):
            raise AssertionError("t did not preserve parity")
        if (p_value & 1) == (state & 1) or (u_value & 1) == (state & 1):
            raise AssertionError("p/u did not flip parity")
        if state & 1:
            if p_value != u_value:
                raise AssertionError("p and u must coincide on odd states")
            if len({t_value, p_value, u_value}) != 2:
                raise AssertionError("odd state must have two distinct children")
        else:
            if len({t_value, p_value, u_value}) != 3:
                raise AssertionError("even state must have three distinct children")
        checked += 1
    return {
        "positive_states_checked": checked,
        "odd_state_children": {"odd": 1, "even": 1, "identity": "p(x)=u(x)"},
        "even_state_children": {"odd": 2, "even": 1},
        "parity_matrix": [[1, 1], [2, 1]],
        "spectral_radius": 1 + math.sqrt(2),
        "all_checks_pass": True,
    }


def reachable_rows(maximum_length: int) -> tuple[dict[str, list[dict[str, int]]], dict[str, set[int]]]:
    pell = pell_numbers(maximum_length)
    rows: dict[str, list[dict[str, int]]] = {}
    cumulative_sets: dict[str, set[int]] = {}
    for phase, initial in PHASE_START.items():
        current = {initial}
        cumulative: set[int] = set()
        phase_rows: list[dict[str, int]] = []
        for length in range(1, maximum_length + 1):
            if length > 1:
                current = {
                    successor
                    for state in current
                    for successor in next_arithmetic_states(state)
                }
            cumulative.update(current)
            even = sum(state & 1 == 0 for state in current)
            odd = len(current) - even
            if len(current) > pell[length]:
                raise AssertionError("exact arithmetic image exceeded Pell bound")
            if len(cumulative) > pell_cumulative_bound(length):
                raise AssertionError("cumulative arithmetic image exceeded Pell sum")
            phase_rows.append(
                {
                    "length": length,
                    "distinct_exact_length_states": len(current),
                    "distinct_cumulative_states": len(cumulative),
                    "even_states": even,
                    "odd_states": odd,
                    "pell_exact_bound": pell[length],
                    "pell_cumulative_bound": pell_cumulative_bound(length),
                }
            )
        rows[phase] = phase_rows
        cumulative_sets[phase] = cumulative
    return rows, cumulative_sets


def fringe_step(state: int) -> int:
    if state < 0:
        raise ValueError("fringe state must be nonnegative")
    packed = 1 + 2 * state
    odd = packed ^ ((packed >> 1) | (packed >> 2))
    return (odd << 1) ^ (odd | (odd >> 1))


def actual_driver(length: int) -> tuple[str, ...]:
    state = 0
    output: list[str] = []
    for _ in range(length):
        output.append("u" if state & 3 == 0 else "t")
        state = fringe_step(state)
    return tuple(output)


def _inverse_t_mod(output: int, width: int) -> int:
    if width < 0:
        raise ValueError("width must be nonnegative")
    if width == 0:
        return 0
    output &= (1 << width) - 1
    state = 0
    for position in range(width):
        previous_one = (state >> (position - 1)) & 1 if position >= 1 else 0
        previous_two = (state >> (position - 2)) & 1 if position >= 2 else 0
        bit = ((output >> position) & 1) ^ (previous_one | previous_two)
        state |= bit << position
    return state


def inverse_generator_mod(name: str, output: int, width: int) -> int:
    if width == 0:
        return 0
    mask = (1 << width) - 1
    output &= mask
    if name == "t":
        return _inverse_t_mod(output, width)
    if name == "u":
        return _inverse_t_mod(output ^ 1, width)
    if name == "p":
        recovered_low_bit = (output & 1) ^ 1
        adjusted = output ^ 1 ^ (2 if recovered_low_bit == 0 else 0)
        return _inverse_t_mod(adjusted, width)
    raise ValueError("unknown generator")


def backward_zero_branch(q_name: str, successor: int, width: int) -> int:
    if q_name not in ("t", "u"):
        raise ValueError("branch must be t or u")
    if width < 2:
        raise ValueError("width must be at least two")
    inner_width = width - 2
    state = inverse_generator_mod(q_name, successor, inner_width)
    state = inverse_generator_mod("p", state, inner_width)
    return ((state << 2) | 3) & ((1 << width) - 1)


def schedule_survivor_residue(q_names: tuple[str, ...], width: int) -> int:
    if width < 2 or width % 2:
        raise ValueError("width must be a positive even integer")
    required = width // 2
    if len(q_names) < required:
        raise ValueError("not enough branch letters")
    state = 0
    precision = 0
    for q_name in reversed(q_names[:required]):
        precision += 2
        state = backward_zero_branch(q_name, state, precision)
    return state


def valuation_difference(left: int, right: int, width: int) -> int:
    difference = (left - right) & ((1 << width) - 1)
    if difference == 0:
        return width
    return (difference & -difference).bit_length() - 1


def actual_agreement_rows(maximum_length: int, survivor_bits: int) -> dict[str, Any]:
    if survivor_bits < 2 * maximum_length + 2:
        raise ValueError("survivor width is too small for the requested word length")
    driver = actual_driver(survivor_bits // 2)
    survivor = schedule_survivor_residue(driver, survivor_bits)
    phase_rows: dict[str, list[dict[str, int]]] = {}
    for phase, initial in PHASE_START.items():
        current = {initial}
        cumulative: set[int] = set()
        rows: list[dict[str, int]] = []
        best_valuation = -1
        best_state = 0
        for length in range(1, maximum_length + 1):
            if length > 1:
                current = {
                    successor
                    for state in current
                    for successor in next_arithmetic_states(state)
                }
            cumulative.update(current)
            for state in cumulative:
                valuation = valuation_difference(state, survivor, survivor_bits)
                if valuation > best_valuation:
                    best_valuation = valuation
                    best_state = state
            rows.append(
                {
                    "maximum_length": length,
                    "distinct_exact_length_states": len(current),
                    "distinct_cumulative_states": len(cumulative),
                    "best_matching_bits": best_valuation,
                    "best_complete_pair_depth": best_valuation // 2,
                    "best_state": best_state,
                }
            )
        phase_rows[phase] = rows
    return {
        "survivor_bits": survivor_bits,
        "actual_driver_prefix": "".join(driver),
        "actual_survivor_residue": survivor,
        "actual_survivor_residue_hex": hex(survivor),
        "by_phase": phase_rows,
        "interpretation": (
            "a phase word of length at most N can sustain exactly floor(v2(x-X)/2) "
            "complete zero blocks, where x is its arithmetic quotient state"
        ),
        "all_checks_pass": True,
    }


def run_campaign(
    maximum_length: int = DEFAULT_MAXIMUM_LENGTH,
    survivor_bits: int = DEFAULT_SURVIVOR_BITS,
) -> dict[str, Any]:
    if not 1 <= maximum_length <= ABSOLUTE_MAXIMUM_LENGTH:
        raise ValueError("maximum length outside campaign cap")
    if survivor_bits % 2 or survivor_bits < 2 * maximum_length + 2:
        raise ValueError("survivor bits must be even and sufficiently large")
    reachable, _ = reachable_rows(maximum_length)
    rate = math.log(2) / math.log(1 + math.sqrt(2))
    payload: dict[str, Any] = {
        "parity_transition": verify_parity_transition(),
        "reachable_arithmetic_states": reachable,
        "actual_agreement": actual_agreement_rows(maximum_length, survivor_bits),
        "exact_conclusions": {
            "pell_bound": (
                "for a fixed phase, exact-length-n arithmetic witness states are at most P_n"
            ),
            "cumulative_pell_bound": (
                "phase-a witness states of length at most N are at most "
                "(P_(N+1)+P_N-1)/2"
            ),
            "improved_counting": (
                "the same bound applies to distinct driver prefixes with kappa_a(q,L)<=N"
            ),
            "almost_sure_rate": (
                "Bernoulli-almost surely liminf kappa_a(q,L)/L is at least "
                "log(2)/log(1+sqrt(2))"
            ),
        },
        "almost_sure_linear_rate": rate,
        "scope_warning": (
            "bounded arithmetic enumeration does not prove divergence on the actual schedule"
        ),
    }
    digest = hashlib.sha256()
    digest.update(b"rule30-period-two-arithmetic-quotient-v1\0")
    digest.update(json.dumps(payload, sort_keys=True, separators=(",", ":")).encode())
    payload["certificate_sha256"] = digest.hexdigest()
    return payload


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--maximum-length", type=int, default=DEFAULT_MAXIMUM_LENGTH)
    parser.add_argument("--survivor-bits", type=int, default=DEFAULT_SURVIVOR_BITS)
    args = parser.parse_args()
    print(json.dumps(run_campaign(args.maximum_length, args.survivor_bits), sort_keys=True))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: experiments/problem1_nonperiodicity/analyze_period_two_arithmetic_quotient.cpp
#include <algorithm>
#include <bit>
#include <cstdint>
#include <iostream>
#include <string>
#include <unordered_set>
#include <vector>

namespace {

uint64_t forward_t(uint64_t state) {
    return state ^ ((state << 1) | (state << 2));
}

uint64_t fringe_step(uint64_t state) {
    const uint64_t packed = 1 + 2 * state;
    const uint64_t odd = packed ^ ((packed >> 1) | (packed >> 2));
    return (odd << 1) ^ (odd | (odd >> 1));
}

std::vector<char> actual_driver(std::size_t length) {
    uint64_t state = 0;
    std::vector<char> result;
    result.reserve(length);
    for (std::size_t index = 0; index < length; ++index) {
        result.push_back((state & 3) == 0 ? 'u' : 't');
        if (index + 1 < length) {
            state = fringe_step(state);
        }
    }
    return result;
}

uint64_t width_mask(unsigned width) {
    return width == 64 ? ~uint64_t{0} : ((uint64_t{1} << width) - 1);
}

uint64_t inverse_t_mod(uint64_t output, unsigned width) {
    output &= width_mask(width);
    uint64_t state = 0;
    for (unsigned position = 0; position < width; ++position) {
        const uint64_t previous_one = position >= 1 ? ((state >> (position - 1)) & 1) : 0;
        const uint64_t previous_two = position >= 2 ? ((state >> (position - 2)) & 1) : 0;
        const uint64_t bit = ((output >> position) & 1) ^ (previous_one | previous_two);
        state |= bit << position;
    }
    return state;
}

uint64_t inverse_generator_mod(char name, uint64_t output, unsigned width) {
    if (width == 0) {
        return 0;
    }
    output &= width_mask(width);
    if (name == 't') {
        return inverse_t_mod(output, width);
    }
    if (name == 'u') {
        return inverse_t_mod(output ^ 1, width);
    }
    const uint64_t recovered_low_bit = (output & 1) ^ 1;
    const uint64_t adjusted = output ^ 1 ^ (recovered_low_bit == 0 ? 2 : 0);
    return inverse_t_mod(adjusted, width);
}

uint64_t backward_zero_branch(char branch, uint64_t successor, unsigned width) {
    const unsigned inner_width = width - 2;
    uint64_t state = inverse_generator_mod(branch, successor, inner_width);
    state = inverse_generator_mod('p', state, inner_width);
    return ((state << 2) | 3) & width_mask(width);
}

uint64_t actual_survivor_64() {
    const auto driver = actual_driver(32);
    uint64_t state = 0;
    unsigned precision = 0;
    for (auto iterator = driver.rbegin(); iterator != driver.rend(); ++iterator) {
        precision += 2;
        state = backward_zero_branch(*iterator, state, precision);
    }
    return state;
}

int valuation_difference(uint64_t left, uint64_t right) {
    const uint64_t difference = left - right;
    return difference == 0 ? 64 : std::countr_zero(difference);
}

void run_phase(char phase, uint64_t initial, int maximum_length, uint64_t survivor) {
    std::vector<uint64_t> current{initial};
    std::unordered_set<uint64_t> cumulative;
    cumulative.reserve(17000000);

    for (int length = 1; length <= maximum_length; ++length) {
        if (length > 1) {
            std::vector<uint64_t> next;
            next.reserve(current.size() * 3);
            for (uint64_t state : current) {
                const uint64_t stepped = forward_t(state);
                next.push_back(stepped);
                next.push_back(stepped ^ 1);
                if ((state & 1) == 0) {
                    next.push_back(stepped ^ 3);
                }
            }
            std::sort(next.begin(), next.end());
            next.erase(std::unique(next.begin(), next.end()), next.end());
            current.swap(next);
        }

        for (uint64_t state : current) {
            cumulative.insert(state);
        }

        int best_valuation = -1;
        uint64_t best_state = 0;
        for (uint64_t state : cumulative) {
            const int valuation = valuation_difference(state, survivor);
            if (valuation > best_valuation) {
                best_valuation = valuation;
                best_state = state;
            }
        }

        std::cout << phase << ' ' << length << ' ' << current.size() << ' '
                  << cumulative.size() << ' ' << best_valuation << ' '
                  << best_valuation / 2 << ' ' << best_state << '\n';
    }
}

}  // namespace

int main(int argc, char** argv) {
    int maximum_length = 26;
    if (argc == 2) {
        maximum_length = std::stoi(argv[1]);
    }
    if (maximum_length < 1 || maximum_length > 26) {
        std::cerr << "maximum length must be between 1 and 26\n";
        return 2;
    }

    const uint64_t survivor = actual_survivor_64();
    if (survivor != 0x7fe13f3088c146c7ULL) {
        std::cerr << "unexpected actual survivor residue\n";
        return 3;
    }

    std::cout << "survivor_hex 0x" << std::hex << survivor << std::dec << '\n';
    std::cout << "phase length exact_states cumulative_states matching_bits complete_pairs best_state\n";
    run_phase('p', 3, maximum_length, survivor);
    run_phase('u', 1, maximum_length, survivor);
    return 0;
}
